Improvements to mortar friction

recuero opened the pull request:
## Reason
- ~Parallel communication of weighted velocities was not working~ Convergence with `NEWTON` showed significantly poorer behavior for some relatively complex cases with many nodes switching frictional states
- Needed a tool to analyze the frictional state of contacting nodes (in contact, sticking, slipping)

## Design
- Create regular nodal auxiliary kernel to inform the user of frictional states

## Impact
- Analysts can discern whether local material is sticking or slipping
- Use of normalized "c" factor for contact constraints. This makes it easier for users to pick a value that's numerically robust. It should be on the order of the expected contact pressure.

(#17495)

recuero committed:
Fix communication of parallel weighted velocities (#17495)

I noticed parallel behavior was apparently worse than serial's for
problems involving a non-trivial number of contacting nodes. This fix
makes such problems have similar behavior in serial and parallel.

recuero committed:
Add frictional status aux kernel (#17495)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/966257/) on 51976fd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20388/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20388/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3f7be38a624c00a59bbc55609f8cf778683f190b`

recuero committed:
Increase ad size required for test

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/966378/) on 9e30d5c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20388/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/966401/) on 9e30d5c wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>3f7be3</th>
      <th colspan=4>#20388 9e30d5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.03%</td>
      <td>89.05%</td>
      <td>+0.02%</td>
      <td>91.89%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2841</td>
      <td>2871</td>
      <td>+30</td>
      <td>34</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>350</td>
      <td>353</td>
      <td>+3</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20388/coverage_diff/contact.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20388/coverage/contact)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20388/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad reviewed (commented):
I'm not quite sure how this PR is addressing some of the bullets listed in the title post. Specifically the correctness of the weighted velocity communication. The change from `pr.first` to `std::get<0>(tup)` and `pr.second` to `std::get<1>(tup)` should not change the functioning of the code at all, e.g. they should yield equivalent results. I feel like this is reflected in the fact that you didn't have to re-gold anything?

recuero commented on modules/contact/src/constraints/ComputeDynamicFrictionalForceLMMechanicalContact.C:
```diff
@@ -289,7 +289,7 @@ ComputeDynamicFrictionalForceLMMechanicalContact::enforceConstraintOnDof3d(
   else
   {
-    const Real epsilon_sqrt = 1.0e-14;
+    const Real epsilon_sqrt = 1.0e-48;
```
Yeah, it is. We are introducing an error by adding this. The goal, as I intended to do it in the first place, was to avoid the MetaPhysicL error. A lower number still avoids this error, so why having it larger.

lindsayad replied:
I'm all for making it smaller. Just curious how you arrived at 48 😄

recuero replied:
The heuristic process was sqrt(1.0e-48) ~ 1.0e-24 which is probably well below the precision required in the contact constraints for a successful run.

lindsayad replied:
works for me

recuero commented:
I noticed difference in behavior between serial and parallel with more complex models (so I was not expecting to regold anything). But, as you point out, I may be missing something here. I'll do some more testing. The root cause may be somewhere else.

lindsayad commented:
The `epsilon_sqrt` change would make a change in floating point results; just not the switch from pair to tuple syntax

lindsayad commented on modules/contact/include/auxkernels/MortarFrictionalStateAux.h:
```diff
@@ -0,0 +1,46 @@
+  virtual Real computeValue() override;
+
+private:
+  /// Tangent along the first direction
+  const MooseArray<Real> * const _tangent_one;
+
+  /// Tangent along the second direction
+  const MooseArray<Real> * const _tangent_two;
+
+  /// Normal contact pressure
+  const MooseArray<Real> * const _contact_pressure;
```
```suggestion
  /// Tangent along the first direction
  const MooseArray<Real> & _tangent_one;

  /// Tangent along the second direction
  const MooseArray<Real> & _tangent_two;

  /// Normal contact pressure
  const MooseArray<Real> & _contact_pressure;
```

lindsayad commented on modules/contact/src/auxkernels/MortarFrictionalStateAux.C:
```diff
@@ -0,0 +1,85 @@
+
+MortarFrictionalStateAux::MortarFrictionalStateAux(const InputParameters & params)
+  : AuxKernel(params),
+    _tangent_one(&coupledValueLower("tangent_one")),
+    _tangent_two(&coupledValueLower("tangent_two")),
+    _contact_pressure(&coupledValueLower("contact_pressure")),
```
```suggestion
    _tangent_one(coupledValueLower("tangent_one")),
    _tangent_two(coupledValueLower("tangent_two")),
    _contact_pressure(coupledValueLower("contact_pressure")),
```

recuero replied:
I enabling use of the mortar aux kernel by 2D problems. For that reason, I am keeping `tangent_two` as a pointer, but maybe there was a better way to do that.

lindsayad reviewed (changes requested):
Other than that, this looks good to me

recuero committed:
Address Alex's comments.

Revert nonsense change on syntax. I did this because I was seeing different parallel behavior
but this was due to poor scaling in a large problem. Results appear to make sense now.

Other changes:

 - Allow frictional mortar state aux kernel to work in 2D too.
 - Allow normalize_c to work with frictional objects (normalization wasn't defined ealier: fixes bad access)

recuero commented:
@loganharbour: I am adding here one example to the `modules examples` contact recipe. This test does not require storage of a mesh since it's using a mesh generator and checking a couple of csv files.

loganharbour commented:
> @loganharbour: I am adding here one example to the `modules examples` contact recipe. This test does not require storage of a mesh since it's using a mesh generator and checking a couple of csv files.

If we're looking at adding single tests that take 1:30, I would prefer that we move them to a scheduled task that we run every week or so. In my opinion this is definitely too long (I was thinking 10-20 minutes per test), and doesn't allow for much room for growth.

What's the memory usage like for this run?

Ping @idaholab/moose-ccb for thoughts on how we want to deal with examples. See https://civet.inl.gov/job/969158/

lindsayad commented on modules/contact/src/auxkernels/MortarFrictionalStateAux.C:
```diff
@@ -0,0 +1,84 @@
+MortarFrictionalStateAux::MortarFrictionalStateAux(const InputParameters & params)
+  : AuxKernel(params),
+    _tangent_one(coupledValueLower("tangent_one")),
+    _tangent_two(isParamValid("tangent_two") ? coupledValueLower("tangent_two") : _zero),
```
Should it be an error if the mesh dimension is 3 and the second tangent is not provided?

recuero replied:
Yeah, thanks. Will update the PR.

lindsayad commented on modules/contact/src/auxkernels/MortarFrictionalStateAux.C:
```diff
@@ -0,0 +1,84 @@
+  if (_contact_pressure[_qp] > _tolerance)
+    status = 2;
+
+  Real tangential_pressure =
```
```suggestion
  const Real tangential_pressure =
```

lindsayad commented on modules/contact/src/auxkernels/MortarFrictionalStateAux.C:
```diff
@@ -0,0 +1,84 @@
+  Real tangential_pressure =
+      std::sqrt(_tangent_one[_qp] * _tangent_one[_qp] + _tangent_two[_qp] * _tangent_two[_qp]);
+
+  Real tangential_pressure_sat = _mu * _contact_pressure[_qp];
```
```suggestion
  const Real tangential_pressure_sat = _mu * _contact_pressure[_qp];
```

recuero committed:
Address review: Add error on mortar frictional status aux kernel

lindsayad commented on modules/contact/src/auxkernels/MortarFrictionalStateAux.C:
```diff
@@ -0,0 +1,91 @@
+    paramError("boundary",
+               "MortarFrictionalStateAux auxiliary kernel must be restricted to a boundary.");
+
+  const auto mortar_dimension = _fe_problem.mesh().dimension() - 1;
```
```suggestion
  const auto mortar_dimension = _subproblem.mesh().dimension() - 1;
```

lindsayad commented on modules/contact/src/auxkernels/MortarFrictionalStateAux.C:
```diff
@@ -0,0 +1,91 @@
+
+MortarFrictionalStateAux::MortarFrictionalStateAux(const InputParameters & params)
+  : AuxKernel(params),
+    _fe_problem(*params.get<FEProblemBase *>("_fe_problem_base")),
```
```suggestion
```

lindsayad commented on modules/contact/include/auxkernels/MortarFrictionalStateAux.h:
```diff
@@ -0,0 +1,49 @@
+  virtual Real computeValue() override;
+
+private:
+  /// Handle to feproblem to find out problem dimension
+  const FEProblemBase & _fe_problem;
+
```
```suggestion
```

lindsayad replied:
```suggestion
```

lindsayad reviewed (changes requested):
should just be able to use the `_subproblem`
